vmm/src/vm.rs

```diff
@@ -390,24 +416,73 @@ impl DeviceManager {
         vm_fd: &Arc<VmFd>,
         vm_cfg: &VmConfig,
         msi_capable: bool,
+        userspace_ioapic: bool,
     ) -> DeviceManagerResult<Self> {
         let io_bus = devices::Bus::new();
         let mut mmio_bus = devices::Bus::new();
 
-        // Add serial device
-        let serial_evt = EventFd::new(EFD_NONBLOCK).map_err(DeviceManagerError::EventFd)?;
-        let serial_irq_fd = serial_evt.as_raw_fd();
-        let serial_irq_cb =
-            Box::new(move || serial_evt.write(1)) as devices::legacy::SerialInterrupt;
+        let ioapic = if userspace_ioapic {
+            let vm_fd_clone = vm_fd.clone();
+            let ioapic_cb = Box::new(move |msg: ioapic::MsiMessage| {
+                let msi = kvm_msi {
+                    address_lo: msg.addr,
+                    address_hi: 0x0,
+                    data: msg.data,
+                    flags: 0u32,
+                    devid: 0u32,
+                    pad: [0u8; 12],
+                };
+
+                match vm_fd_clone.signal_msi(msi) {
+                    Ok(ret) => {
+                        if ret > 0 {
+                            debug!("MSI message successfully delivered");
+                            Ok(())
+                        } else {
+                            warn!("failed to deliver MSI message, blocked by guest");
+                            Err(ioapic::Error::InterruptGuestBlocked)
+                        }
+                    }
+                    Err(e) => Err(ioapic::Error::InterruptFailed(e)),
+                }
+            }) as ioapic::IoapicInterrupt;
+
+            // Create IOAPIC
+            Some(Arc::new(Mutex::new(ioapic::Ioapic::new(ioapic_cb))))
+        } else {
+            None
+        };
+
+        let interrupt_info = InterruptInfo {
+            msi_capable,
+            ioapic: &ioapic,
+        };
+
+        let serial_irq_cb = if let Some(ioapic) = &ioapic {
+            let ioapic_clone = ioapic.clone();
+            Box::new(move || {
+                ioapic_clone.lock().unwrap().service_irq(4).map_err(|e| {
+                    std::io::Error::new(
+                        std::io::ErrorKind::Other,
+                        format!("failed to inject IRQ #4: {:?}", e),
+                    )
+                })
+            }) as devices::legacy::SerialInterrupt
+        } else {
+            let serial_evt = EventFd::new(EFD_NONBLOCK).map_err(DeviceManagerError::EventFd)?;
+            // Serial is tied to IRQ #4
+            vm_fd
+                .register_irqfd(serial_evt.as_raw_fd(), 4)
+                .map_err(DeviceManagerError::Irq)?;
+
+            Box::new(move || serial_evt.write(1)) as devices::legacy::SerialInterrupt
+        };
```

Comment:
Again, the above code can be made generic, other than be bound to serial.

Reply:
Yes I will pull directly from `devices`, instead of `devices::legacy::Serial`

Reply:
I think what @chao-p means is that this routine could be a generic one, given an `Option<ioapic>` and an IRQ number, for all non MSI based devices. It could be defined in `devices/src/lib.rs` for example, and reused here.

Reply:
Yes I moved the definition to `devices/src/lib.rs`. Regarding the addition of the parameters Option<ioapic> and IRQ number, it will not buy us anything here, since the pin based case does not use the IRQ number anyway.